// @gmod/cram/errors.js
"use strict";

var _getPrototypeOf = require("babel-runtime/core-js/object/get-prototype-of");

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require("babel-runtime/helpers/possibleConstructorReturn");

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require("babel-runtime/helpers/inherits");

var _inherits3 = _interopRequireDefault(_inherits2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CramError = function (_Error) {
  (0, _inherits3.default)(CramError, _Error);

  function CramError() {
    (0, _classCallCheck3.default)(this, CramError);
    return (0, _possibleConstructorReturn3.default)(this, (CramError.__proto__ || (0, _getPrototypeOf2.default)(CramError)).apply(this, arguments));
  }

  return CramError;
}(Error);

/** Error caused by encountering a part of the CRAM spec that has not yet been implemented */


var CramUnimplementedError = function (_Error2) {
  (0, _inherits3.default)(CramUnimplementedError, _Error2);

  function CramUnimplementedError() {
    (0, _classCallCheck3.default)(this, CramUnimplementedError);
    return (0, _possibleConstructorReturn3.default)(this, (CramUnimplementedError.__proto__ || (0, _getPrototypeOf2.default)(CramUnimplementedError)).apply(this, arguments));
  }

  return CramUnimplementedError;
}(Error);

/** An error caused by malformed data.  */


var CramMalformedError = function (_CramError) {
  (0, _inherits3.default)(CramMalformedError, _CramError);

  function CramMalformedError() {
    (0, _classCallCheck3.default)(this, CramMalformedError);
    return (0, _possibleConstructorReturn3.default)(this, (CramMalformedError.__proto__ || (0, _getPrototypeOf2.default)(CramMalformedError)).apply(this, arguments));
  }

  return CramMalformedError;
}(CramError);

/**
 * An error caused by attempting to read beyond the end of the defined data.
 */


var CramBufferOverrunError = function (_CramMalformedError) {
  (0, _inherits3.default)(CramBufferOverrunError, _CramMalformedError);

  function CramBufferOverrunError() {
    (0, _classCallCheck3.default)(this, CramBufferOverrunError);
    return (0, _possibleConstructorReturn3.default)(this, (CramBufferOverrunError.__proto__ || (0, _getPrototypeOf2.default)(CramBufferOverrunError)).apply(this, arguments));
  }

  return CramBufferOverrunError;
}(CramMalformedError);

/**
 * An error caused by data being too big, exceeding a size limit.
 */


var CramSizeLimitError = function (_CramError2) {
  (0, _inherits3.default)(CramSizeLimitError, _CramError2);

  function CramSizeLimitError() {
    (0, _classCallCheck3.default)(this, CramSizeLimitError);
    return (0, _possibleConstructorReturn3.default)(this, (CramSizeLimitError.__proto__ || (0, _getPrototypeOf2.default)(CramSizeLimitError)).apply(this, arguments));
  }

  return CramSizeLimitError;
}(CramError);

/**
 * An invalid argument was supplied to a cram-js method or object.
 */


var CramArgumentError = function (_CramError3) {
  (0, _inherits3.default)(CramArgumentError, _CramError3);

  function CramArgumentError() {
    (0, _classCallCheck3.default)(this, CramArgumentError);
    return (0, _possibleConstructorReturn3.default)(this, (CramArgumentError.__proto__ || (0, _getPrototypeOf2.default)(CramArgumentError)).apply(this, arguments));
  }

  return CramArgumentError;
}(CramError);

module.exports = {
  CramBufferOverrunError: CramBufferOverrunError,
  CramMalformedError: CramMalformedError,
  CramUnimplementedError: CramUnimplementedError,
  CramSizeLimitError: CramSizeLimitError,
  CramArgumentError: CramArgumentError
};
