// @gmod/cram/dist/io/bufferCache.js
'use strict';

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LRU = require('lru-cache');

var BufferCache = function () {
  function BufferCache(_ref) {
    var fetch = _ref.fetch,
        _ref$size = _ref.size,
        size = _ref$size === undefined ? 10000000 : _ref$size,
        _ref$chunkSize = _ref.chunkSize,
        chunkSize = _ref$chunkSize === undefined ? 32768 : _ref$chunkSize;
    (0, _classCallCheck3.default)(this, BufferCache);

    if (!fetch) throw new Error('fetch function required');
    this.fetch = fetch;
    this.chunkSize = chunkSize;
    this.lruCache = LRU({ max: Math.floor(size / chunkSize) });
  }

  (0, _createClass3.default)(BufferCache, [{
    key: 'get',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(outputBuffer, offset, length, position) {
        var _this = this;

        var firstChunk, lastChunk, fetches, _loop, chunk, chunks, chunksOffset;

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!(outputBuffer.length < offset + length)) {
                  _context.next = 2;
                  break;
                }

                throw new Error('output buffer not big enough for request');

              case 2:

                // calculate the list of chunks involved in this fetch
                firstChunk = Math.floor(position / this.chunkSize);
                lastChunk = Math.floor((position + length) / this.chunkSize);

                // fetch them all as necessary

                fetches = new Array(lastChunk - firstChunk + 1);

                _loop = function _loop(chunk) {
                  fetches[chunk - firstChunk] = _this._getChunk(chunk).then(function (data) {
                    return {
                      data: data,
                      chunkNumber: chunk
                    };
                  });
                };

                for (chunk = firstChunk; chunk <= lastChunk; chunk += 1) {
                  _loop(chunk);
                }

                // stitch together the response buffer using them
                _context.next = 9;
                return _promise2.default.all(fetches);

              case 9:
                chunks = _context.sent;
                chunksOffset = position - chunks[0].chunkNumber * this.chunkSize;

                chunks.forEach(function (_ref3) {
                  var data = _ref3.data,
                      chunkNumber = _ref3.chunkNumber;

                  var chunkPositionStart = chunkNumber * _this.chunkSize;
                  var copyStart = 0;
                  var copyEnd = _this.chunkSize;
                  var copyOffset = offset + (chunkNumber - firstChunk) * _this.chunkSize - chunksOffset;

                  if (chunkNumber === firstChunk) {
                    copyOffset = offset;
                    copyStart = chunksOffset;
                  }
                  if (chunkNumber === lastChunk) {
                    copyEnd = position + length - chunkPositionStart;
                  }

                  data.copy(outputBuffer, copyOffset, copyStart, copyEnd);
                });

              case 12:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function get(_x, _x2, _x3, _x4) {
        return _ref2.apply(this, arguments);
      }

      return get;
    }()
  }, {
    key: '_getChunk',
    value: function _getChunk(chunkNumber) {
      var cachedPromise = this.lruCache.get(chunkNumber);
      if (cachedPromise) return cachedPromise;

      var freshPromise = this.fetch(chunkNumber * this.chunkSize, this.chunkSize);
      this.lruCache.set(chunkNumber, freshPromise);
      return freshPromise;
    }
  }]);
  return BufferCache;
}();

module.exports = BufferCache;
